File: iosApp/iosApp/Modules/DegenCult/CultProposal/CultProposalWireframe.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultCultProposalWireframe {
    private weak var parent: UIViewController?
    private let context: CultProposalWireframeContext
    
    private weak var vc: UIViewController?

    init(
        _ parent: UIViewController?,
        context: CultProposalWireframeContext
    ) {
        self.parent = parent
        self.context = context
    }
}

extension DefaultCultProposalWireframe {

    func present() {
        let vc = wireUp()
        parent?.show(vc, sender: self)
    }

    func navigate(to destination: CultProposalWireframeDestination) {
        if destination is CultProposalWireframeDestination.Dismiss {
            vc?.popOrDismiss()
        }
    }
}

private extension DefaultCultProposalWireframe {

    func wireUp() -> UIViewController {
        let vc: CultProposalViewController = UIStoryboard(.cultProposal).instantiate()
        let presenter = DefaultCultProposalPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            context: context
        )
        vc.presenter = presenter
        self.vc = vc
        return vc
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/cultproposal/CultProposalWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.cultproposal

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.NavigationFragment
import com.sonsofcrypto.web3walletcore.modules.degenCultProposal.CultProposalWireframe
import com.sonsofcrypto.web3walletcore.modules.degenCultProposal.CultProposalWireframeContext
import com.sonsofcrypto.web3walletcore.modules.degenCultProposal.CultProposalWireframeDestination
import com.sonsofcrypto.web3walletcore.modules.degenCultProposal.DefaultCultProposalPresenter

class DefaultCultProposalWireframe(
    private val parent: WeakRef<Fragment>?,
    private val context: CultProposalWireframeContext,
): CultProposalWireframe {

    override fun present() {
        val fragment = wireUp()
        (parent?.get() as? NavigationFragment)?.push(fragment, true)
    }

    override fun navigate(destination: CultProposalWireframeDestination) {
        when (destination) {
            is CultProposalWireframeDestination.Dismiss -> {
                (parent?.get() as? NavigationFragment)?.popFragment()
            }
        }
    }

    private fun wireUp(): Fragment {
        val view = CultProposalFragment()
        val presenter = DefaultCultProposalPresenter(
            WeakRef(view),
            this,
            context,
        )
        view.presenter = presenter
        return view
    }
}